src/paginator.js

```diff
@@ -0,0 +1,300 @@
+import assign from 'lodash.assign';
+import omit from 'lodash.omit';
+
+
+export const styles = Object.freeze({
+  pageNumber: 1,
+  limitOffset: 2,
+});
+
+export const errors = Object.freeze({
+  invalidStyle: 'Invalid pagination style',
+  calculateOffsetNoLimit: 'Can\'t calculate offsets without a limit',
+  calculatePageNoLimit: 'Can\'t calculate pages without a limit'
+});
+
+const getResponseBody = function(response) {
+  return response.json();
+};
+
+const defaultOptions = {
+  limitQueryParam: 'limit',
+  offsetQueryParam: 'offset',
+  pageQueryParam: 'page',
+  style: styles.pageNumber
+};
+
+const defaultQueryParams = {
+  limit: null,
+  offset: 0,
+  page: 1
+};
+
+export class Paginator {
+  constructor(request, requestOptions = null, queryParams = {}, options = {}) {
+    this._cache = Object.create(null);
+    this._parsedResponse = null;
+    this._request = request;
+    this._requestOptions = assign({}, requestOptions);
+
+    this._setOptions(options);
+    this._setQueryParams(queryParams);
+  }
+
+  get page() {
+    return this._page;
+  }
+
+  next() {
+    const incrementPage = (pageCount) => {
+      if (this._page < pageCount) {
+        this._page++;
+      }
+    };
+
+    return this.fetchPageCount()
```

Comment:
I think this is a race condition. You should be incrementing the page before making the request, else two calls to `next` will return the same page.

I think you'll need to decrement the `page` during instantiation and then this will increment up to the desired page. As a result, `first` should set the page to `0`